Add more tests to the autograd wrt view and inplace

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31147 Add more tests to the autograd wrt view and inplace**

The goal here is to add more tests of the current behavior of the autograd to make sure no regressions are introduced when modifying it.
Do let me know if you think of other corner cases I missed.

Differential Revision: [D19301082](https://our.internmc.facebook.com/intern/diff/D19301082)

albanD committed:
Add more tests to the autograd wrt view and inplace

[ghstack-poisoned]

albanD committed:
Update on "Add more tests to the autograd wrt view and inplace"




[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 6258c79:
* **1/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=6258c79855fa97525a631961d435e416cfac27cc).

### 1 failure *not* recognized by patterns:
| Job | Step |
| --- | --- |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>pytorch_linux_xenial_py3_clang5_mobile_build</sup> | <sup>Build</sup> |


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

albanD committed:
Update on "Add more tests to the autograd wrt view and inplace"


The goal here is to add more tests of the current behavior of the autograd to make sure no regressions are introduced when modifying it.
Do let me know if you think of other corner cases I missed.

[ghstack-poisoned]

albanD committed:
Update on "Add more tests to the autograd wrt view and inplace"


The goal here is to add more tests of the current behavior of the autograd to make sure no regressions are introduced when modifying it.
Do let me know if you think of other corner cases I missed.

[ghstack-poisoned]

ezyang commented on test/test_autograd.py:
```diff
@@ -3333,6 +3333,318 @@ def test_checkpointing(self):
+    def test_autograd_views_codegen(self):
+        # This is not necessarily the absolute correct behavior, but this is the current
+        # one. This test is here to make sure that any change to this behavior is detected
+        # and not silent.
```
Maybe, explain what acceptable other behaviors are? I.e. what are the degrees of freedom in the test?

albanD replied:
Changing any of these test will be a BC-breaking change. All these tests use only public api so users could be using it.

ezyang commented on test/test_autograd.py:
```diff
@@ -3333,6 +3333,318 @@ def test_checkpointing(self):
+        # should_raise contains None if it should not raise
+        # should_raise contains a string of the error if it should raise
+        # The 3 elements are for view_as, first output of unbind and second output of unbind
+        run_test(True, False, True, (None, None, None))
```
These tests would be easier to read with kwargs, IMO.

albanD committed:
Update on "Add more tests to the autograd wrt view and inplace"


The goal here is to add more tests of the current behavior of the autograd to make sure no regressions are introduced when modifying it.
Do let me know if you think of other corner cases I missed.

[ghstack-poisoned]

albanD committed:
Update on "Add more tests to the autograd wrt view and inplace"


The goal here is to add more tests of the current behavior of the autograd to make sure no regressions are introduced when modifying it.
Do let me know if you think of other corner cases I missed.

[ghstack-poisoned]

gchanan commented on test/test_autograd.py:
```diff
@@ -3333,6 +3333,343 @@ def test_checkpointing(self):
+        # and not silent. The TODOs below mark the places with unexpected behavior.
+        # Note that any change in these test will be BC-breaking and should be done carefully.
+
+        # This test checks the behavior that sample codegen functions (view_as and unbind)
```
I'm not sure what "sample codegen functions" means?

albanD replied:
Subset of functions from the codegen selected by hand. I'll say that explicitly.

gchanan commented on test/test_autograd.py:
```diff
@@ -3333,6 +3333,343 @@ def test_checkpointing(self):
+                    return inp.unbind()
+
+            def maybe_check_raise(fn, should_raise):
+                if should_raise is not None:
```
super nit: prefer "if should_raise:" -- it would be surprising if someone passed False to should_raise and it was interpreted as True.

albanD replied:
I prefer None vs string to False vs string.
I'll add an assert that it is either None or a string to make sure this does not happen.

gchanan commented on test/test_autograd.py:
```diff
@@ -3333,6 +3333,343 @@ def test_checkpointing(self):
+        # The 3 elements are for view_as, first output of unbind and second output of unbind
+        run_test(grad_mode=True, requires_grad=False, is_view=True,
+                 should_raise_tuple=(None, None, None))
+        # TODO: Second should_raise should not be None below
```
interesting -- the second return of unbind is different than the first?

albanD replied:
The view is only checked at runtime and the first output being a view is allowed.
The second output is not though.

But these cases should not even exist: we should never contruct such Node. This is (should be) enforced when constructing Nodes for an autograd.Function. But the codegen does it by hand so no other check happen at construction time.

gchanan commented on test/test_autograd.py:
```diff
@@ -3333,6 +3333,343 @@ def test_checkpointing(self):
+                 should_raise_tuple=(None, None, "INTERNAL ASSERT FAILED"))
+        # TODO: views require gradients when created in no_grad mode but their grad_fn is not populated
+        run_test(grad_mode=False, requires_grad=True, is_view=True,
+                 should_raise_tuple=("leaf Variable that requires grad",
```
nit: I'd just put the full error message -- I spent awhile trying to guess what the actual error was.

albanD replied:
I wanted to reduce line size. But I'll get a common string with the full error message.

albanD committed:
Update on "Add more tests to the autograd wrt view and inplace"


The goal here is to add more tests of the current behavior of the autograd to make sure no regressions are introduced when modifying it.
Do let me know if you think of other corner cases I missed.

[ghstack-poisoned]

ezyang commented on test/test_autograd.py:
```diff
@@ -3333,6 +3333,343 @@ def test_checkpointing(self):
+                 should_raise_tuple=(None, None, None))
+        # TODO: Second should_raise should not be None below
+        run_test(grad_mode=True, requires_grad=True, is_view=True,
+                 should_raise_tuple=(None, None, "diff_view_meta->output_nr_ == 0 INTERNAL ASSERT FAILED"))
```
The contract of INTERNAL ASSERT is that it should be impossible to trigger it from userland (aka Python). So this is just a straight up bug; either the assert is incorrectly classified, or we're doing something wrong here

albanD replied:
Yes, both the 2nd and 3rd here should raise an error stating that we're doing an inplace on the output of a function where it's not allowed.

ezyang commented on test/test_autograd.py:
```diff
@@ -3333,6 +3333,343 @@ def test_checkpointing(self):
+            self.assertTrue(out[1]._is_view() == is_view)
+
+            # Are inplace allowed?
+            out = get_view_as()
```
I don't understand, why do you need to allocate fresh variables for the inplace test? The code here would be a lot more direct if you just inlined the `view_as` and `get_unbind` implementations; they seem to be factored into functions so you can call them twice but the second call seems completely unnecessary.

albanD replied:
Yes I'll clean this up.

ezyang commented on test/test_autograd.py:
```diff
@@ -3333,6 +3333,343 @@ def test_checkpointing(self):
+        # This test checks the behavior of two codegen functions (view_as and unbind)
+        # with respect to view tracking and inplace operation on the output.
+
+        def run_test(grad_mode, requires_grad, is_view, should_raise_tuple):
```
The ordering of the arguments here is a little backward. First you allocate the variable (using `requires_grad`), then you run the operation with grad enabled or disabled (`grad_mode`). But in the argument list here they're reverse order

ezyang replied:
In every usage of this function, `is_view` is true. I don't think we're ever going to make `view_as` or `unbind` return not views? So why have them be arguments to this function?

albanD replied:
Well I would argue that we should not return views in `torch.no_grad()` mode.
Because here we "autodiff" the view op even though it was in a no grad block.
See https://github.com/pytorch/pytorch/issues/26546 for reference of this problem.

ezyang replied:
oooooh this is giving me flashbacks to another issue. This one: https://github.com/pytorch/pytorch/pull/12502 (and also #11390)

ezyang commented on test/test_autograd.py:
```diff
@@ -3333,6 +3333,343 @@ def test_checkpointing(self):
+
+        # I) Simple views
+        bw_called = [0]
+        ga_nz = [False]
```
I think it would be very helpful for readers if it was explained what these indicator variables represent

ezyang commented on test/test_autograd.py:
```diff
@@ -3333,6 +3333,343 @@ def test_checkpointing(self):
+                    self.assertTrue(bw_called[0] == expected_called)
+                    self.assertTrue(ga_nz[0] == expected_ga_nz)
+
+        # II) Complex views
```
Split this into a separate test

albanD committed:
Update on "Add more tests to the autograd wrt view and inplace"


The goal here is to add more tests of the current behavior of the autograd to make sure no regressions are introduced when modifying it.
Do let me know if you think of other corner cases I missed.

[ghstack-poisoned]

albanD committed:
Update on "Add more tests to the autograd wrt view and inplace"


The goal here is to add more tests of the current behavior of the autograd to make sure no regressions are introduced when modifying it.
Do let me know if you think of other corner cases I missed.

[ghstack-poisoned]

albanD committed:
Update on "Add more tests to the autograd wrt view and inplace"


The goal here is to add more tests of the current behavior of the autograd to make sure no regressions are introduced when modifying it.
Do let me know if you think of other corner cases I missed.

[ghstack-poisoned]

albanD committed:
Update on "Add more tests to the autograd wrt view and inplace"


The goal here is to add more tests of the current behavior of the autograd to make sure no regressions are introduced when modifying it.
Do let me know if you think of other corner cases I missed.

[ghstack-poisoned]

albanD committed:
Update on "Add more tests to the autograd wrt view and inplace"


The goal here is to add more tests of the current behavior of the autograd to make sure no regressions are introduced when modifying it.
Do let me know if you think of other corner cases I missed.

Differential Revision: [D19301082](https://our.internmc.facebook.com/intern/diff/D19301082)

[ghstack-poisoned]
